src/lib/x509/pkix_enums.h

```diff
@@ -100,18 +102,76 @@ BOTAN_PUBLIC_API(2,0) const char* to_string(Certificate_Status_Code code);
 * X.509v3 Key Constraints.
 * If updating update copy in ffi.h
 */
-enum Key_Constraints {
-   NO_CONSTRAINTS     = 0,
-   DIGITAL_SIGNATURE  = 1 << 15,
-   NON_REPUDIATION    = 1 << 14,
-   KEY_ENCIPHERMENT   = 1 << 13,
-   DATA_ENCIPHERMENT  = 1 << 12,
-   KEY_AGREEMENT      = 1 << 11,
-   KEY_CERT_SIGN      = 1 << 10,
-   CRL_SIGN           = 1 << 9,
-   ENCIPHER_ONLY      = 1 << 8,
-   DECIPHER_ONLY      = 1 << 7
-};
+class BOTAN_PUBLIC_API(3,0) Key_Constraints
+   {
+   public:
+      enum Key_Constraints_Bits : uint32_t {
+         None              = 0,
+         DigitalSignature  = 1 << 15,
+         NonRepudiation    = 1 << 14,
+         KeyEncipherment   = 1 << 13,
+         DataEncipherment  = 1 << 12,
+         KeyAgreement      = 1 << 11,
+         KeyCertSign       = 1 << 10,
+         CrlSign           = 1 << 9,
+         EncipherOnly      = 1 << 8,
+         DecipherOnly      = 1 << 7,
+
+         // Deprecated SHOUTING_CASE names for Key_Constraints
+         // will be removed in a future major release
+         NO_CONSTRAINTS    = None,
+         DIGITAL_SIGNATURE = DigitalSignature,
+         NON_REPUDIATION   = NonRepudiation,
+         KEY_ENCIPHERMENT  = KeyEncipherment,
+         DATA_ENCIPHERMENT = DataEncipherment,
+         KEY_AGREEMENT     = KeyAgreement,
+         KEY_CERT_SIGN     = KeyCertSign,
+         CRL_SIGN          = CrlSign,
+         ENCIPHER_ONLY     = EncipherOnly,
+         DECIPHER_ONLY     = DecipherOnly,
+      };
+
+      Key_Constraints(const Key_Constraints& other) = default;
+      Key_Constraints(Key_Constraints&& other) = default;
+      Key_Constraints& operator=(const Key_Constraints& other) = default;
+      Key_Constraints& operator=(Key_Constraints&& other) = default;
+
+      Key_Constraints(Key_Constraints_Bits bits) : m_value(bits) {}
+
+      explicit Key_Constraints(uint32_t bits) : m_value(bits) {}
+
+      Key_Constraints() : m_value(0) {}
+
+      bool operator==(const Key_Constraints& other) const { return m_value == other.m_value; }
+      bool operator!=(const Key_Constraints& other) const { return m_value != other.m_value; }
+
+      bool operator==(Key_Constraints_Bits other) const { return m_value == other; }
+      bool operator!=(Key_Constraints_Bits other) const { return m_value != other; }
```

Comment:
C++20 fanciness:

```suggestion
      bool operator==(const Key_Constraints&) const = default;
```

* [Default comparisons](https://en.cppreference.com/w/cpp/language/default_comparisons)
* [Operator rewriting](https://en.cppreference.com/w/cpp/language/overload_resolution#Call_to_an_overloaded_operator)

The operators taking `Key_Constraints_Bits` can be omitted due to implicit conversion via `Key_Constraints(Key_Constraint_Bits)`.

Reply:
Very nice thank you